pkg/cmd/server/start/start_allinone.go

```diff
@@ -209,6 +218,35 @@ func (o *AllInOneOptions) Complete() error {
 	o.NodeArgs.NodeName = strings.ToLower(o.NodeArgs.NodeName)
 	o.NodeArgs.MasterCertDir = o.MasterOptions.MasterArgs.ConfigDir.Value()
 
+	// For backward compatibility of DNS queries to the master service IP, enabling node DNS
+	// continues to start the master DNS, but the container DNS server will be the node's.
+	// However, if the user has provided an override DNSAddr, we need to honor the value if
+	// the port is not 53 and we do that by disabling node DNS.
+	if !o.IsRunFromConfig() && o.NodeArgs.Components.Enabled(ComponentDNS) {
+		dnsAddr := &o.MasterOptions.MasterArgs.DNSBindAddr
+
+		if dnsAddr.Provided {
+			if dnsAddr.Port == 53 {
+				// the user has set the DNS port to 53, which is the effective default (node on 53, master on 8053)
+				dnsAddr.Port = 8053
+				dnsAddr.URL.Host = net.JoinHostPort(dnsAddr.Host, strconv.Itoa(dnsAddr.Port))
+			} else {
+				// if the user set the DNS port to anything but 53, disable node DNS since ClusterDNS (and glibc)
+				// can't look up DNS on anything other than 53, so we'll continue to use the proxy.
+				o.NodeArgs.Components.Disable(ComponentDNS)
```

Comment:
Hrm.  Ugh.

On Fri, May 27, 2016 at 1:13 PM, Jordan Liggitt <notifications@github.com>
wrote:

> In pkg/cmd/server/start/start_allinone.go
> <https://github.com/openshift/origin/pull/7807#discussion_r64937727>:
>
> > +	// For backward compatibility of DNS queries to the master service IP, enabling node DNS
> > +	// continues to start the master DNS, but the container DNS server will be the node's.
> > +	// However, if the user has provided an override DNSAddr, we need to honor the value if
> > +	// the port is not 53 and we do that by disabling node DNS.
> > +	if !o.IsRunFromConfig() && o.NodeArgs.Components.Enabled(ComponentDNS) {
> > +		dnsAddr := &o.MasterOptions.MasterArgs.DNSBindAddr
> > +
> > +		if dnsAddr.Provided {
> > +			if dnsAddr.Port == 53 {
> > +				// the user has set the DNS port to 53, which is the effective default (node on 53, master on 8053)
> > +				dnsAddr.Port = 8053
> > +				dnsAddr.URL.Host = net.JoinHostPort(dnsAddr.Host, strconv.Itoa(dnsAddr.Port))
> > +			} else {
> > +				// if the user set the DNS port to anything but 53, disable node DNS since ClusterDNS (and glibc)
> > +				// can't look up DNS on anything other than 53, so we'll continue to use the proxy.
> > +				o.NodeArgs.Components.Disable(ComponentDNS)
>
> shouldn't we warn them here if they explicitly enabled DNS on the node?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin/pull/7807/files/58ddd8589d00920f33047cd23685716381e1b859#r64937727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ABG_p6ky4s7lAbJjS64yaM4FNxGvj7L7ks5qFyYtgaJpZM4Hp34y>
> .
>